### ISSUE TITLE
preserving spir-v developer trust via careful conversion between SPIR-V and WGSL

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#692]
There have been concerns in the SPIR-V community about the decision that WebGPU will use WGSL and only WGSL.
In the early days after the decision, I spoke in hand-wavy ways about bijectivity between SPIR-V and WGSL.
Those ideas were discussed somewhat in #582 and #620 

I've tried to better capture developer feedback and motivation and further refine the ideas, in this [slide deck](https://docs.google.com/presentation/d/1ybmnmmzsZZTA0Y9awTDUNqkkMo3a2N7wR4Z04SQinPg/edit?usp=sharing)   also attached in PDF: 
[Preserving developer trust via careful SPIR-V WGSL conversions 2020-04-06.pdf](https://github.com/gpuweb/gpuweb/files/4446559/Preserving.developer.trust.via.careful.SPIR-V.WGSL.conversions.2020-04-06.pdf)

This was discussed at length in the shader language meeting on 2020-04-07: [minutes](https://docs.google.com/document/d/1SVCmA9-fLYNBJwePlsWXZhBHmmrKHuBJStBlPcgARj0/edit#bookmark=id.niyvrcrluswx)

**Comment by kvark** [comment]
Thank you for the presentation as well as filing this!
The story from my perspective:
  1. we agreed on SPIR-V semantics in a text human-writable shape
  2. many users have SPIR-V based shader toolchains, the ecosystem there is pretty well established with a wide range of tools
  3. we don't have a good criteria of what should go into WGSL, what shouldn't, and what should be sugared over

What @dneto0 proposed is a framework that naturally builds on top of (1) to solve the need of (3) and satisfy use case (2) by definition. The solution is a workflow framework based on the concept of canonization of SPIR-V. It would allow us to reason about sugar and features we consider putting into WGSL and make steady progress.

**Comment by litherum** [comment]
I started a public email thread about this [here](https://lists.w3.org/Archives/Public/public-gpu/2020Mar/0024.html). I'm happy to move the discussion to GitHub if that is better. 

(I intentionally chose email rather than a GitHub issue because this discussion isn't going to result in any specific change to the spec, and isn't about any technical proposal or feature. This is more of a meta issue. But if people want to do this on GitHub, I'll comment here about it.)

**Comment by dneto0** [comment]
>  I'm happy to move the discussion to GitHub if that is better.

I much prefer a GitHub issue to email because there's a much better chance that we converge.  Email conversations are a DAG and can be very hard to follow.

An issue doesn't have to have a direct impact on the spec.  The issue text and its resolution serves as institutional memory: it's a record for future folks to understand decisions and directions.

**Comment by litherum** [comment]
We agree with many of the ideas presented here. Preserving developer trust is valuable. Cross-compiling shaders from a different source language to WGSL is an important use case. Developers shouldn’t have to fight against drivers. We should enable developers to create great content, without them being surprised about performance or behavior. Extremely complicated compiler transformations are error-prone. Compiling WGSL to SPIR-V should be behavior- and performance-preserving, just as it should be for compiling WGSL to MSL and WGSL to HLSL. Determining a language canonicalization function would provide insight. These are all things to strive for. (This is an incomplete list; there are also plenty of other things to strive for, like raw performance, portability across multiple OSes, security, etc.)

Some potential language modifications exist which clearly have no effect on these goals, like re-ordering SPIR-V IDs or switching the SPIR-V source order of the then/else block in a conditional. Other potential modifications clearly _would_ have an effect on these goals, like if all shader invocations were required to be executed in serial by a single compute thread. The standardization group needs to determine some criteria by which we can determine if a proposed change affects the goals too much or not.

### Criteria

“Understandability" is one criterion. It’s not a bad criterion; it does provide signal. Unfortunately, it’s qualitative, and is dramatically affected by who you ask. We should go further, and have evidence-based criteria. A complex change with huge runtime benefits would be more valuable than a simple change with little runtime benefits. Users are above authors, who are above implementors, in the W3C’s [priority of constituencies](https://dev.w3.org/html5/html-design-principles/#priority-of-constituencies).

It’s impossible to prove a negative, that a proposed change has no behavior or correctness problems on any devices in the world, so the burden of proof must be on the party claiming performance/behavior problems to demonstrate it. If a performance/behavior problem is demonstrated, that is sufficient to constrain the design of WGSL. But if there’s no evidence provided, then there’s no reason to constrain the language’s design against beneficial proposals.

The way we inspire developers to trust us is to make well-informed, data-driven, and well-documented design decisions, coupled with a development focus of shipping bug-free and fully-featured implementations. If a particular transformation doesn’t work well on a popular driver on a popular device, it’s totally valid to incorporate that evidence into the design of WGSL. But it isn’t valid to constrain the design of WGSL because of an unreliable metric gathered from a non-representative group of developers.

### Proposal

With that in mind, here is a proposed test for the inclusion of a design constraint to WGSL due to unexpected performance/correctness problems in particular APIs and drivers:

- A class of shaders exists, written in a human-writable shading language, which a human might have reasonably written
- … when compiled from that human-writable language into the backend language using the latest version of a popular compiler, and then the construct is perturbed (e.g. a variable hoisted, or a const promoted to a variable, or duplicate code moved to a common function, etc.),
- … it results in a reproducible and significant performance or correctness problem
- … on a shipping system configuration of the most recent version of a major browser (because only new browser versions will get WebGPU)
- … and the shipping configuration has a nontrivial number of users (otherwise we could simply disallow WebGPU on those drivers)

If this can be demonstrated, then that is sufficient to constrain the design of WGSL. This applies not just to SPIR-V, but also to HLSL and MSL: If a particular language construct can be demonstrated to cause behavior/correctness problems when compiling to those languages, then that evidence is also sufficient to constrain WGSL.

**Comment by krogovin** [comment]
The story that has been brought up over multiple issues that there will be tools to compile/convert to WGSL. As a developer, it would go a very long way if these tools were part of an official tool chain, much like the SPIR-V toolchain that Khronos provides and sponsors.

Going further, since one of WGSL's goals is to be semi-interchangeable with SPIR-V, such a tool chain should include the following:
- both a command line and library interface (ideally realized in a fashion to make it easily callable from WASM modules) to convert from SPIR-V to WGSL
- a set of guidelines on how to compile to SPIR-V so that the tool above produces performant shaders; there are loads of optimization passes available in SPIR-V tools (and each pass can be run multiple times).

As I've stated in #677, the syntax for WGSL puts me in the place of either needing to rewrite shaders (which is a severe sin) or to have a tool chain to covert my existing shaders (possibly with mild tweaking) to something consumable by WebGPU.

That the shading language is different makes it a hard requirement to have a sponsored, useable tool chain to deal with WGSL for cross-platform shaders.

**Comment by dj2** [comment]
That's the plan for [Tint](https://dawn.googlesource.com/tint), I also believe [Naga](https://github.com/gfx-rs/naga) will enable those flows as well.

**Comment by dneto0** [comment]
While I have an example in front of me, we just got this bug report from a user:

https://github.com/KhronosGroup/SPIRV-Tools/issues/3460 

It turns out SPIRV-Cross was sinking quite expensive operations into a loop, causing a serious performance degradation.

It's an example which I think meets @litherum's criteria for something to be avoided, but does *not* have implications for WGSL design.  I consider it a bug in SPIRV-Cross, to be fixed.

**Comment by krogovin** [comment]
I think that bug report should be viewed as more than just a one-off issue. The real issue is that by relying on a large tool-chain to do change OpLoad/OpStore to SSA equivalent along with potentially oodles of other optimizations is quite risky. I strongly suspect that the plan to use SPIRV-Cross to do that work is going to make more and more issues and risk the entire stack.

This is looking to me like a symptom that the goal "convertible trivially to/from SPIR-V" but be human readable/writable is not really tractable. Perhaps, it would be wiser in Vulkan platforms to convert WGSL to a higher level language that the SPIRV shader compiler accepts and then use SPIR-V opt to optimize it; given that on Direct3D and Metal, the conversion has to be a high level language anyways, this would make the spirit the same across platforms better.

**Comment by Kangz** [comment]
@dneto0 can we close this issue?

**Comment by dneto0** [comment]
Closed per discussion in the virtual f2f
